### PR TITLE
feat: add appeal duration options to policy

### PR DESCRIPTION
**Pull request by bsushmith**
Fixes https://github.com/odpf/guardian/issues/213

Related Proton PR - https://github.com/odpf/proton/pull/166

* Adds appeal duration configuration options to policy
* For permanent access, the value of duration is `0h`
* Adds a new API - `GET policies/:id/version/:version/preferences`

**Note**:
This change does not touch the appeal config in the provider. That needs to be tackled separately. Issue: https://github.com/odpf/guardian/issues/226

### CODE REVIEW

**Comment by bsushmith** [comment]
Am having issues fixing the two failed repository test cases due to the sql orm. Marking the PR as a draft till that is resolved.

File: api/handler/v1beta1/adapter.go [hunk 1]
```diff
@@ -393,6 +407,24 @@ func (a *adapter) ToPolicyProto(p *domain.Policy) (*guardianv1beta1.Policy, erro
 	return policyProto, nil
 }
 
+func (a *adapter) ToPolicyAppealConfigProto(p *domain.Policy) *guardianv1beta1.PolicyAppealConfig {
```
Comment (rahmatrhd) [inline api/handler/v1beta1/adapter.go:410]:
```suggestion
func (a *adapter) ToPolicyAppealConfigProto(p *domain.Policy) *guardianv1beta1.PolicyAppealConfig {
	if p.Appeal == nil {
		return nil
	}
```

File: domain/appeal.go [hunk 2]
```diff
@@ -75,7 +75,7 @@ func (a *Appeal) Cancel() {
 func (a *Appeal) Activate() error {
 	a.Status = AppealStatusActive
 
-	if a.Options != nil && a.Options.Duration != "" {
+	if a.Options != nil && a.Options.Duration != "" && a.Options.Duration != "0h" {
```
Comment (rahmatrhd) [inline domain/appeal.go:78]: let's use `time.Duration` to check if the duration is zero

**Review by rahmatrhd — CHANGES_REQUESTED** [review]

File: api/handler/v1beta1/adapter.go (hunk 1 above)
Comment (bsushmith) [reply api/handler/v1beta1/adapter.go:410]: This will not work as `p.Appeal` is not a slice, it is a struct.

File: domain/appeal.go (hunk 2 above)
Comment (bsushmith) [reply domain/appeal.go:79]: Done. modified the logic accordingly

Comment (bsushmith) [reply domain/appeal.go]: also updated the documentation for valid duration values.

File: domain/policy.go [hunk 3]
```diff
@@ -182,15 +182,30 @@ type Requirement struct {
...
-	Requirements []*Requirement    `json:"requirements,omitempty" yaml:"requirements,omitempty" validate:"omitempty,min=1,dive"`
-	Labels       map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
-	IAM          *IAMConfig        `json:"iam,omitempty" yaml:"iam,omitempty" validate:"omitempty,dive"`
-	CreatedAt    time.Time         `json:"created_at,omitempty" yaml:"created_at,omitempty"`
-	UpdatedAt    time.Time         `json:"updated_at,omitempty" yaml:"updated_at,omitempty"`
+	ID           string             `json:"id" yaml:"id" validate:"required"`
+	Version      uint               `json:"version" yaml:"version" validate:"required"`
+	Description  string             `json:"description" yaml:"description"`
+	Steps        []*Step            `json:"steps" yaml:"steps" validate:"required,min=1,dive"`
+	Appeal       PolicyAppealConfig `json:"appeal" yaml:"appeal" validate:"omitempty"`
```
Comment (rahmatrhd) [inline domain/policy.go:189]: should this be `*PolicyAppealConfig`? because the existing policies are not having this field

Comment (bsushmith) [reply domain/policy.go:189]: The `gorm` migration will autofill the table with a default value for this column when the column gets added.
But given that migration will be done by `go-migrate` in the latest master. Will check this out

[the remark this answers is not in the document]
Comment (bsushmith) [reply domain/policy.go:189]: makes sense. can't recall why I went with this. but making changes to rename the field as `AppealConfig` in the domain and also in the repository model.

[the remark this answers is not in the document]
Comment (bsushmith) [reply domain/policy.go]: Done.

File: api/handler/v1beta1/adapter.go [hunk 4]
```diff
@@ -393,6 +407,24 @@ func (a *adapter) ToPolicyProto(p *domain.Policy) (*guardianv1beta1.Policy, erro
...
+
+	var durationOptions []*guardianv1beta1.PolicyAppealConfig_DurationOptions
+	if p.Appeal.DurationOptions != nil {
+		var option *guardianv1beta1.PolicyAppealConfig_DurationOptions
+		for _, d := range p.Appeal.DurationOptions {
+			option = &guardianv1beta1.PolicyAppealConfig_DurationOptions{
+				Name:  d.Name,
+				Value: d.Value,
+			}
+			durationOptions = append(durationOptions, option)
```
[the remark this answers is not in the document]
Comment (bsushmith) [reply api/handler/v1beta1/adapter.go]: done

File: domain/policy.go (hunk 3 above)
Comment (bsushmith) [reply domain/policy.go]: Done. made the suggested change.

File: api/handler/v1beta1/adapter.go (hunk 1 above)
Comment (bsushmith) [reply api/handler/v1beta1/adapter.go:412]: now that `p.AppealConfig` is a pointer, added this check.

File: domain/policy.go [hunk 5]
```diff
@@ -182,15 +182,30 @@ type Requirement struct {
...
-	Requirements []*Requirement    `json:"requirements,omitempty" yaml:"requirements,omitempty" validate:"omitempty,min=1,dive"`
-	Labels       map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
-	IAM          *IAMConfig        `json:"iam,omitempty" yaml:"iam,omitempty" validate:"omitempty,dive"`
-	CreatedAt    time.Time         `json:"created_at,omitempty" yaml:"created_at,omitempty"`
-	UpdatedAt    time.Time         `json:"updated_at,omitempty" yaml:"updated_at,omitempty"`
+	ID           string              `json:"id" yaml:"id" validate:"required"`
+	Version      uint                `json:"version" yaml:"version" validate:"required"`
+	Description  string              `json:"description" yaml:"description"`
+	Steps        []*Step             `json:"steps" yaml:"steps" validate:"required,min=1,dive"`
+	AppealConfig *PolicyAppealConfig `json:"appeal_config" yaml:"appeal_config" validate:"omitempty"`
```
Comment (rahmatrhd) [inline domain/policy.go:189]:
```suggestion
	AppealConfig *PolicyAppealConfig `json:"appeal_config" yaml:"appeal_config" validate:"omitempty,dive"`
```
CMIIW but I think `dive` is needed if we want to validate AppealConfig struct as well when validating the Policy struct.

**Review by rahmatrhd — APPROVED** [review]
lgtm

**Review by singhvikash11 — APPROVED** [review]

**Pull request merged**